internal/test/test.go

```diff
@@ -0,0 +1,53 @@
+package test
+
+import (
+	"context"
+	"reflect"
+	"testing"
+)
+
+type Test interface {
+	Run(context.Context, *testing.T)
+}
+
+type test struct {
+	cs     []Caser
+	target func(ctx context.Context, c Caser) ([]interface{}, error)
+}
+
+func New(opts ...Option) Test {
+	t := new(test)
+	for _, opt := range append(defaultOptions, opts...) {
+		opt(t)
+	}
+	return t
+}
+
+func (test *test) Run(ctx context.Context, t *testing.T) {
+	t.Helper()
+	for _, c := range test.cs {
+		t.Run(c.Name(), func(tt *testing.T) {
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			gots, err := test.target(ctx, c)
+			if err != nil {
+				tt.Error(err)
+			}
+
+			if len(c.Wants()) != len(gots) {
+				tt.Fatalf("wants and gots length are not equals. wants: %d, gots: %d", len(c.Wants()), len(gots))
+			}
+
+			for i, want := range c.Wants() {
+				if reflect.DeepEqual(want, gots[i]) {
```

Comment:
in some case we want to check the specific variable in the struct rather than the full struct... (like having a timestamp inside the struct, or random value, etc.) 
I think this implementation cannot support this.

Reply:
Thank you for your review.
You're right.
So, we make "Checkfunc", and Is it better for users to cast ?

Reply:
Yes, I think it is better :)

Reply:
Thank you！
I will fix it soon.

Reply:
what do you think if we change the logic to "if checkfunc is nil then use the deepequal method".

Reply:
Thank you for your review.
I think it's good!
Since I modified the example, so could you please confirm it?

Reply:
I think it's good :) thank you!!

Reply:
I added `fieldFunc`, becase we should validate parameter of mock.
Do you think this??
Cloud you please read example??

Reply:
Sorry I can't understand the use case of it, can you describe more?

Reply:
Sorry. when we want to set mock object, I should validate parameter.
The following is example.


```go
// mock object
type MockRepository struct {
	RegisterFunc func(id int, data string) error
}

func (m *MockRepository) Register(id int, data string) {
	return m.RegisterFunc(id, data)
}


// test case
type test struct {
	fieldFunc func(*testing.T) Field
}

tests := []test {
	func() test{
		var (
			wantId = 11111
			wantTxt = "teext"
		)

		return test {
			args: args {
				id wantId,
				txt: wantTxt,
			},
			fieldFunc: func(t *testing.T) Field {
				t.Helper()

				mockR := &MockRepository {
					RegisterFunc: func(id int, data string) {
						if id != wantId {
							t.Errorf("argument is not equals.")
						}

						if data != wantTxt {
							t.Errorf("argument is not equals.")
						}
					}
				}

				return Field {
					repo: mockR,
				}
			}
		}
	}
}


```

Reply:
I think we can set the field like this... what do you think?
```golang
tests := []test {
	func() test{
		var (
			wantId = 11111
			wantTxt = "teext"
		)

		return test {
			args: args {
				id wantId,
				txt: wantTxt,
			},
			WithField: func() Field {
				mockR := &MockRepository {
					RegisterFunc: func(id int, data string) {
						if id != wantId {
							t.Errorf("argument is not equals.")
						}

						if data != wantTxt {
							t.Errorf("argument is not equals.")
						}
					}
				}

				return Field {
					repo: mockR,
				}
			}(), // call the function and return  the field
		}
	}
}
```

Reply:
`t.Errorf` is root `tesing.T` object.
When an error is output, it is output not as a subtest error but as a root error.
we want output as subtest.

Reply:
I think it is good :)

Reply:
Thank you for review :bow: